File: Python/백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
import sys
input = sys.stdin.readline

from collections import deque

def dfs(V): # V: 시작 정점

    stack = [V]
    visited = []

    while stack:
        n = stack.pop()

        if n not in visited:
            visited.append(n)

            for i in sorted(graph[n], reverse=True):
                if i not in visited:
                    stack.append(i)

    print(" ".join(map(str, visited)))

def bfs(V): # V: 시작 정점

    q = deque([V])
    visited = []

    while q:
        n = q.popleft()

        if n not in visited:
            visited.append(n)

        for i in sorted(graph[n]):
            if i not in visited:
                q.append(i)

    print(" ".join(map(str, visited)))

N, M, V = map(int, input().split()) # N: 정점 수, M: 간선 수, V: 시작 정점

graph = [[] for _ in range(N+1)]

for _ in range(M):
    u, v = map(int, input().split())
    graph[u].append(v)
    graph[v].append(u)


# 중복제거
for i in range(1, N+1):
    graph[i] = list(set(graph[i]))

dfs(V)
bfs(V)

File: Java/백준/Silver/1260. DFS와 BFS/DFS와 BFS.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main{
  static List<Integer>[] graph;
  static boolean[] visited;
  public static void main(String[] args) throws Exception{
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    StringTokenizer st = new StringTokenizer(br.readLine());
    int n = Integer.parseInt(st.nextToken()); // 정점 수
    int m = Integer.parseInt(st.nextToken()); // 간선 수
    int v = Integer.parseInt(st.nextToken()); // 시작노드
    graph = new ArrayList[n + 1];
    for (int i = 1; i <= n; i++) {
      graph[i] = new ArrayList<>();
    }

    for (int i = 0; i < m; i++) {
      StringTokenizer line = new StringTokenizer(br.readLine());
      int s = Integer.parseInt(line.nextToken());
      int e = Integer.parseInt(line.nextToken());
      graph[s].add(e);
      graph[e].add(s);
    }

    for (int i = 1; i <= n; i++) {
      Collections.sort(graph[i]);
    }

    visited = new boolean[n+1];
    dfs(v);
    System.out.println();
    visited = new boolean[n+1];
    bfs(v);

  }

  static void dfs(int node) {
    visited[node] = true;
    System.out.print(node + " ");

    for(int next : graph[node]) {
      if (!visited[next]) {
        dfs(next);
      }
    }
  }

  static void bfs(int start) {
    Queue<Integer> q = new LinkedList<>();
    visited[start] = true;
    q.offer(start);

    while(!q.isEmpty()) {
      int node = q.poll();
      System.out.print(node + " ");

      for(int next : graph[node]) {
        if (!visited[next]) {
          visited[next] = true;
          q.offer(next);
        }
      }
    }
  }

}